src/vm-vcpu/src/vm.rs

```diff
@@ -44,29 +55,59 @@ pub type Result<T> = std::result::Result<T, Error>;
 
 impl KvmVm {
     /// Create a new `KvmVm`.
-    pub fn new(kvm: &Kvm) -> Result<Self> {
-        let fd = kvm.create_vm().map_err(Error::CreateVm)?;
-        Ok(KvmVm {
-            fd,
+    pub fn new<M: GuestMemory>(kvm: &Kvm, vm_state: VmState, guest_memory: &M) -> Result<Self> {
+        let vm_fd = kvm.create_vm().map_err(Error::CreateVm)?;
+
+        let vm = KvmVm {
+            state: vm_state,
+            fd: vm_fd,
             vcpus: Vec::new(),
             vcpu_handles: Vec::new(),
-        })
+        };
+
+        vm.configure_memory_regions(guest_memory, kvm)?;
+
+        mptable::setup_mptable(guest_memory, vm.state.num_vcpus).map_err(Error::Mptable)?;
+
+        vm.setup_irq_controller()?;
```

Comment:
Does setting up the irq controller here creates issues on Arm (I think there was that discussion that vCPUs have to be created first on that platform). If so, maybe one way around this going forward is to create just the raw vcpu fds to begin with for Arm, and set up the proper vcpu wrappers later.

Reply:
I was considering taking a different approach with aarch64 actually. Most of the calls in this function do not make sense for aarch64, but I was thinking of reusing function names, and just call them at the appropriate time depending on the platform. I think I can better explain this in the code directly. For now I would leave the setup here, and on aarch64 the irq setup should transparently be called when creating the vcpus.

Reply:
Sure we can clarify this later.